### PR TITLE
ACS-4025: Support include count and orderBy count for GET /tags

### PR DESCRIPTION
**Pull request by gevagel**
This PR introduces changes to support `?orderBy=count` when the count is included for `GET /tags` API endpoint. The `TagsImpl` method logic has been changed in order to follow the structure suggested in a comment in the associated JIRA issue [link](https://alfresco.atlassian.net/browse/ACS-4025).

### CODE REVIEW

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/tags/GetTagsTests.java [hunk 1]
```diff
@@ -72,6 +77,130 @@ public void getTagsWithConsumerRole()
...
 
+    /**
+     * Include count in the query parameters and ensure count is as expected for returned tags.
+     */
+    @Test(groups = { TestGroup.REST_API, TestGroup.TAGS, TestGroup.REGRESSION })
+    public void testGetTags_withIncludeCount()
+    {
+        STEP("Get tags including count filter and ensure count is as expected for returned tags");
+        returnedCollection = restClient.authenticateUser(adminUserModel)
+                .withParams("include=count")
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/tags/GetTagsTests.java:88]: You can use `include("count")` instead of this if you like.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/tags/GetTagsTests.java [hunk 2]
```diff
@@ -72,6 +77,130 @@ public void getTagsWithConsumerRole()
...
+    /**
+     * Get tags and order results by count. Default sort order should be ascending
+     */
+    @Test(groups = { TestGroup.REST_API, TestGroup.TAGS, TestGroup.REGRESSION })
+    public void testGetTags_withOrderByCountDefaultOrderShouldBeAsc()
+    {
+
+        STEP("Get tags and order results by count. Default sort order should be ascending");
+        returnedCollection = restClient.authenticateUser(adminUserModel)
+                .withParams("include=count&orderBy=count")
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/tags/GetTagsTests.java:111]: Out of interest what happens if count is not included, but we request to sort by it? I'm not sure if we need a test for this, but we should document the behaviour in the API explorer.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/tags/GetTagsTests.java [hunk 3]
```diff
@@ -194,7 +323,7 @@ public void useHighSkipCountCheckPagination()
                 .and().field("hasMoreItems").is("false")
                 .and().field("count").is("0")
                 .and().field("skipCount").is(20000)
-                .and().field("totalItems").is(0);
+                .and().field("totalItems").is(4);
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/tags/GetTagsTests.java:326]: Maybe even better to delete this line of the assertion?

**Review by azakrzewski-hy — APPROVED** [review]
LGTM

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/tags/GetTagsTests.java (hunk 2 above)
Comment (gevagel) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/tags/GetTagsTests.java:111]: The way I wrote the code, we can only sort by count when count is also included. That being said, do you think that when count isn't included and we request to sort by it, we should let the user know why it isn't possible?

Comment (tpage-alfresco) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/tags/GetTagsTests.java:111]: That would be ideal. This is an edge case, but it would be good to document the behaviour and provide a helpful error message.

File: repository/src/main/java/org/alfresco/service/cmr/tagging/TaggingService.java [hunk 4]
```diff
@@ -88,10 +89,7 @@
      * @return PagingResults
      */
     @NotAuditable
-    default PagingResults<Pair<NodeRef, String>> getTags(StoreRef storeRef, PagingRequest pagingRequest, Collection<String> exactNamesFilter, Collection<String> alikeNamesFilter)
-    {
-        return new EmptyPagingResults<>();
-    }
+    List<Pair<NodeRef, String>> getTags(StoreRef storeRef, PagingRequest pagingRequest, Collection<String> exactNamesFilter, Collection<String> alikeNamesFilter);
```
Comment (mpichura) [inline repository/src/main/java/org/alfresco/service/cmr/tagging/TaggingService.java:92]: See my comment for above method

File: repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java [hunk 5]
```diff
@@ -122,7 +122,7 @@ public enum Depth {IMMEDIATE, ANY};
      * @param sortByName boolean
      */
     @Auditable(parameters = {"storeRef", "aspectName", "pagingRequest", "sortByName"})
-    PagingResults<ChildAssociationRef> getRootCategories(StoreRef storeRef, QName aspectName, PagingRequest pagingRequest, boolean sortByName);
+    Collection<ChildAssociationRef> getRootCategories(StoreRef storeRef, QName aspectName, PagingRequest pagingRequest, boolean sortByName);
```
Comment (mpichura) [inline repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java:125]: As I commented for `TaggingService` - I would strongly prefer adding methods than changing their signatures.

File: repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java [hunk 6]
```diff
@@ -135,7 +135,7 @@ public enum Depth {IMMEDIATE, ANY};
      * @return
      */
     @Auditable(parameters = {"storeRef", "aspectName", "pagingRequest", "sortByName", "filter"})
-    PagingResults<ChildAssociationRef> getRootCategories(StoreRef storeRef, QName aspectName, PagingRequest pagingRequest, boolean sortByName, String filter);
+    Collection<ChildAssociationRef> getRootCategories(StoreRef storeRef, QName aspectName, PagingRequest pagingRequest, boolean sortByName, String filter);
```
Comment (mpichura) [inline repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java:138]: See my comment for the above method.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/TagsImpl.java [hunk 7]
```diff
@@ -158,36 +161,33 @@ public void deleteTagById(StoreRef storeRef, String tagId) {
...
-            {
-                for (Pair<String, Integer> tagByCountElem : tagsByCount)
-                {
-                    tagsByCountMap.put(tagByCountElem.getFirst(), tagByCountElem.getSecond());
-                }
-            }
-            tags.forEach(tag -> tag.setCount(Optional.ofNullable(tagsByCountMap.get(tag.getTag())).orElse(0)));
-        }
+		Map<Integer, Collection<String>> namesFilters = resolveTagNamesQuery(params.getQuery());
+		List<Pair<NodeRef, String>> results = taggingService.getTags(storeRef, Util.getPagingRequest(paging), namesFilters.get(EQUALS), namesFilters.get(MATCHES));
```
[the remark this answers is not in the document]
Comment (mpichura) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/TagsImpl.java:171]: To some extent I agree with Tom - the sorting must be done before paging. 
The question is whether it should be done inside this method or service method is not very relevant to me as in both cases we seem do be doing  the sorting in memory on the whole result set. 
So it's either getting all the tags (call the service method to get all tags for `StoreRef`) here and then sort and page them or deal with that on the service method side.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/TagsImpl.java [hunk 8]
```diff
@@ -158,36 +161,33 @@ public void deleteTagById(StoreRef storeRef, String tagId) {
 		taggingService.deleteTag(storeRef, tagValue);
 	}
 
-    @Override
+	@Override
```
[the remark this answers is not in the document]
Comment (tpage-alfresco) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/TagsImpl.java:164]: At some point we should reformat this entire file to get rid of the tabs. It's a nightmare in every PR :sweat_smile:

File: remote-api/src/main/java/org/alfresco/rest/api/impl/TagsImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (tpage-alfresco) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/TagsImpl.java:171]: We could do it in memory after the service call, but I thought that if we do it inside the service call then we're one step closer to replacing it with a DB query that handles these things without loading all tags in the system. I'm happy with either approach as long as the paging actually works with custom sort orders.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/TagsImpl.java [hunk 9]
```diff
@@ -335,4 +336,25 @@ private NodeRef checkTagRootAsNodePrimaryParent(String tagId, NodeRef tagNodeRef
 		}
 		return tagNodeRef;
 	}
+
+	private List<Tag> sortTags(List<Tag> tagsList, String sortBy, boolean sortAsc)
```
Comment (mpichura) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/TagsImpl.java:340]: Is this method ever used?

File: remote-api/src/test/java/org/alfresco/rest/api/impl/TagsImplTest.java [hunk 10]
```diff
@@ -121,13 +127,14 @@ public void setup()
     public void testGetTags()
     {
         given(parametersMock.getPaging()).willReturn(pagingMock);
-        given(taggingServiceMock.getTags(any(StoreRef.class), any(PagingRequest.class), any(), any())).willReturn(pagingResultsMock);
-        given(pagingResultsMock.getTotalResultCount()).willReturn(new Pair<>(Integer.MAX_VALUE, 0));
-        given(pagingResultsMock.getPage()).willReturn(List.of(new Pair<>(TAG_NODE_REF, TAG_NAME)));
+        given(parametersMock.getSorting()).willReturn(new ArrayList<>());
```
Comment (mpichura) [inline remote-api/src/test/java/org/alfresco/rest/api/impl/TagsImplTest.java:130]: I think `Collections.emptyList()` would be more meaningful here and in all other places you've used `(new ArrayList<>()` to mock `include` and `sorting`.

File: repository/src/main/java/org/alfresco/service/cmr/tagging/TaggingService.java [hunk 11]
```diff
@@ -327,8 +330,15 @@ default PagingResults<Pair<NodeRef, String>> getTags(StoreRef storeRef, PagingRe
...
-    
-    
+
+    /**
+     *
+     * @param storeRef              node reference
+     * @param parametersInclude     the 'include' part of the parameters
+     * @return List of pairs of tag nodeRefs and tag names
+     */
+    List<Pair<NodeRef, String>> getTags(StoreRef storeRef, List<String> parametersInclude);
```
Comment (mpichura) [inline repository/src/main/java/org/alfresco/service/cmr/tagging/TaggingService.java:340]: Same here regarding possible use of `Map` instead of `List<Pair>`

File: repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java [hunk 12]
```diff
@@ -150,11 +150,19 @@ public enum Depth {IMMEDIATE, ANY};
...
+                                                              Collection<String> exactNamesFilter, Collection<String> alikeNamesFilter) {
         return new EmptyPagingResults<>();
     }
 
+    /**
+     * Get a collection of the root categories for an aspect/classification supporting multiple name filters.
+     */
+    @Auditable(parameters = {"storeRef", "aspectName", "exactNamesFilter", "alikeNamesFilter"})
+    default Collection<ChildAssociationRef> getRootCategories(StoreRef storeRef, QName aspectName, Collection<String> exactNamesFilter, Collection<String> alikeNamesFilter) {
+        return null;
```
Comment (mpichura) [inline repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java:162]: It woluld be better if `emptyList` (or `emptySet` but I see that the impementation returns `List`) was returned here instead of `null`.

**Review by mpichura — COMMENTED** [review]
I had some questions and remarks. 
Also, as an aside - please check curly brackets formatting in some places.

**Comment by tpage-alfresco** [comment]
Late to the party, but on the topic of modifying the interfaces then I agree with Maciej about TaggingService and CategoryService as these are both marked `@AlfrescoPublicApi`. For interfaces without this annotation then we should feel free to modify the interface as we feel is best.

**Review by MateuszHY — APPROVED** [review]

**Review by krdabrowski — APPROVED** [review]

File: repository/src/main/java/org/alfresco/service/cmr/tagging/TaggingService.java [hunk 13]
```diff
@@ -94,6 +95,18 @@ default PagingResults<Pair<NodeRef, String>> getTags(StoreRef storeRef, PagingRe
     {
         return new EmptyPagingResults<>();
     }
+
+    /**
+     * Get a map of tags filtered by name and sorted by tag name or count
```
Comment (mpichura) [inline repository/src/main/java/org/alfresco/service/cmr/tagging/TaggingService.java:100]: Minor one - can be fixed in another iteration - javadoc here should explain what key and values are (either in general method description on in `@return` part.

File: repository/src/main/java/org/alfresco/service/cmr/tagging/TaggingService.java [hunk 14]
```diff
@@ -362,6 +374,13 @@ default List<Pair<String, NodeRef>> createTags(StoreRef storeRef, List<String> t
     {
         return Collections.emptyList();
     }
+
+    /**
+     *
+     * @param storeRef
+     * @return a map with each tag name and its count
```
Comment (mpichura) [inline repository/src/main/java/org/alfresco/service/cmr/tagging/TaggingService.java:381]: Minor - can come in another iteration.
```suggestion
     * @return a map with each tag name and its usage count

```

**Review by mpichura — APPROVED** [review]
Had some very minor comments - they may be resolved in another iteration.
